### PR TITLE
Feature al 195 update statistics while algo is running

### PR DESCRIPTION
**Pull request by mjoksa**

### CODE REVIEW

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Entities/StatisticsSummaryEntity.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Enumerators;
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Entities
+{
+    public class StatisticsSummaryEntity : TableEntity
+    {
+        public string InstanceId { get; set; }
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Entities/StatisticsSummaryEntity.cs:9]: We can extract properties that will be used in both entities in one main - Base entity (BaseStatisticEntity) or interface.
InstanceId, InstanceType, AlgoInstanceTypeValue  are the same

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Entities/StatisticsEntity.cs [hunk 2]
```diff
@@ -15,5 +15,17 @@ public class StatisticsEntity : TableEntity
         public double? Amount { get; set; }
 
         public bool? IsStarted { get; set; }
+
+        public string AlgoInstanceTypeValue { get; set; }
+
+        public AlgoInstanceType InstanceType
+        {
+            get
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Entities/StatisticsEntity.cs]: Why do we need Algo Instance type?

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs [hunk 3]
```diff
@@ -13,39 +14,82 @@ namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories
...
+
+            if (data.IsStarted.HasValue && data.IsStarted.Value)
+                entitySummary.TotalNumberOfStarts++;
+
+            if (data.IsBuy.HasValue)
+            {
+                entitySummary.TotalNumberOfTrades++;
+
+                if(!entity.Amount.HasValue)
+                    throw new ArgumentException("Amount for statistics not provided");
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs]: Will this throw of exception log the error into CSharpAlgoTemplateLog db table?

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs [hunk 4]
```diff
@@ -13,39 +14,82 @@ namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories
...
+
+                if (!entity.Price.HasValue)
+                    throw new ArgumentException("Price for statistics not provided");
+
+                if (data.IsBuy.Value)
+                {
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs]: This should be different depending on which asset we trade. I also think this values should be calculated before we call the repo. When we have the life version we should use the trading service and then extract the real values of assets after trading.

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Services/WorkflowService.cs [hunk 5]
```diff
@@ -89,8 +89,10 @@ public Task StartAsync()
             var quoteGeneration = _quoteProviderService.Initialize();
             _quoteProviderService.Subscribe(_algo.AssetPair, OnQuote);
             _quoteProviderService.Start();
+            
             //Update algo statistics
-            _statisticsService.OnAlgoStarted();
+            //TODO: Get asset one and two balance and get initial wallet balance before statistics is started
+            _statisticsService.OnAlgoStarted(10000, 5000, 5000);
```
Comment (iliyan-ivanov-prime) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Services/WorkflowService.cs]: Not sure where we are going to get these arguments from ?

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs [hunk 6]
```diff
@@ -13,39 +14,82 @@ namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories
...
+                    entitySummary.AssetOneBalance = entitySummary.AssetOneBalance + entity.Amount.Value;
+                    entitySummary.AssetTwoBalance = entitySummary.AssetTwoBalance - (entity.Amount.Value * entity.Price.Value);
+                    entitySummary.LastWalletBalance = entitySummary.AssetOneBalance + entitySummary.AssetTwoBalance;
```
Comment (iliyan-ivanov-prime) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs]: I think these calculations are not all corrent. Also, for live instances we will get the LastWalletBalance from an external service. These calculations should be used only for demo and backtesting I think.

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Models/StatisticsSummary.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+        public double InitialWalletBalance { get; set; }
+
+        public double LastWalletBalance { get; set; }
+
+        public double AssetOneBalance { get; set; }
+
+        public double AssetTwoBalance { get; set; }
+
+        public double NetProfit => System.Math.Abs(InitialWalletBalance) > 0 ? (InitialWalletBalance - LastWalletBalance) / InitialWalletBalance : 0;
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Models/StatisticsSummary.cs:23]: I think NetProfit will be calculated every time. When we have the "Live" type algo instance, of the wallet will be changed depending on the price - which can be each second, minute. So let's stay this way for the "back-test" but for "Live" I think we should calculate it every time.

File: src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs [hunk 8]
```diff
@@ -13,28 +14,50 @@ namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories
     public class StatisticsRepository : IStatisticsRepository
     {
         private readonly INoSQLTableStorage<StatisticsEntity> _table;
+        private readonly INoSQLTableStorage<StatisticsSummaryEntity> _tableSummary;
 
-        public static readonly string TableName = "Statistics";
+        public static readonly string TableName = "AlgoStatistics";
```
Comment (dafinkagerdjikova) [inline src/Lykke.AlgoStore.CSharp.AlgoTemplate.Models/Repositories/StatisticsRepository.cs]: Statistics will be about Algo Instance, so maybe we should have AlgoInstanceStatistics? it's just a thought.

**Review by dafinkagerdjikova — APPROVED** [review]

**Review by iliyan-ivanov-prime — APPROVED** [review]

**Pull request merged**